lib/g_searcher/search_results/search_result_parser.ex

```diff
@@ -0,0 +1,83 @@
+defmodule GSearcher.SearchResults.SearchResultParser do
+  @top_ads "#tads a"
+  @bottom_ads "#bottomads #tadsb a"
+  @search_results "#search .g div div > a"
+  @total_count "#result-stats"
+  @url_link "href"
+
+  @google_url "https://www.google.com/"
+
+  def parse(all_html) do
+    with {:ok, html_tree} <- Floki.parse_document(all_html),
+         {:ok, top_advertiser_urls} <- fetch_top_advertiser_urls(html_tree),
+         {:ok, regular_advertiser_urls} <- fetch_regular_advertiser_urls(html_tree),
+         {:ok, search_result_urls} <- fetch_search_result_urls(html_tree),
+         {:ok, total_number_of_results} <- fetch_total_result_count(html_tree) do
+      {:ok,
+       %{
+         number_of_results_on_page: Enum.count(search_result_urls),
+         number_of_top_advertisers: Enum.count(top_advertiser_urls),
+         number_of_regular_advertisers: Enum.count(regular_advertiser_urls),
+         search_result_urls: search_result_urls,
+         top_advertiser_urls: top_advertiser_urls,
+         regular_advertiser_urls: regular_advertiser_urls,
+         total_number_of_results: total_number_of_results,
+         html_cache: all_html
+       }}
+    else
+      _ ->
+        {:error, :failed_to_parse_html}
```

Comment:
I'm not sure about this implementation because if 1 of each step is failed then all steps after that step never got executed which is not really productive, what do you think @bterone?

We could change the implementation to process all step, and display the error message if some of them is failed

Reply:
It's primarily for `Floki` when it parses the HTML, I've updated the error handling and bubbled it up to the worker so we can save the error 😄